codegen/src/api/storage.rs

```diff
@@ -61,147 +62,231 @@ fn generate_storage_entry_fns(
     let storage_name = storage_entry.name();
     let Some(storage_hash) = pallet.storage_hash(storage_name) else {
         return Err(CodegenError::MissingStorageMetadata(pallet_name.into(), storage_name.into()));
     };
 
-    let fn_name = format_ident!("{}", storage_entry.name().to_snake_case());
+    let snake_case_name = storage_entry.name().to_snake_case();
     let storage_entry_ty = match storage_entry.entry_type() {
         StorageEntryType::Plain(ty) => *ty,
         StorageEntryType::Map { value_ty, .. } => *value_ty,
     };
     let storage_entry_value_ty = type_gen.resolve_type_path(storage_entry_ty);
-
     let docs = storage_entry.docs();
     let docs = should_gen_docs
         .then_some(quote! { #( #[doc = #docs ] )* })
         .unwrap_or_default();
 
-    let key_args = fields.iter().map(|(field_name, field_type)| {
-        // The field type is translated from `std::vec::Vec<T>` to `[T]`. We apply
-        // Borrow to all types, so this just makes it a little more ergonomic.
-        //
-        // TODO [jsdw]: Support mappings like `String -> str` too for better borrow
-        // ergonomics.
-        let field_ty = match field_type.vec_type_param() {
-            Some(ty) => quote!([#ty]),
-            _ => quote!(#field_type),
-        };
-        quote!( #field_name: impl ::std::borrow::Borrow<#field_ty> )
-    });
-
-    let is_map_type = matches!(storage_entry.entry_type(), StorageEntryType::Map { .. });
-
-    // Is the entry iterable?
-    let is_iterable_type = if is_map_type {
-        quote!(#crate_path::storage::address::Yes)
-    } else {
-        quote!(())
-    };
-
-    let has_default_value = match storage_entry.modifier() {
-        StorageEntryModifier::Default => true,
-        StorageEntryModifier::Optional => false,
+    let is_defaultable_type = match storage_entry.modifier() {
+        StorageEntryModifier::Default => quote!(#crate_path::storage::address::Yes),
+        StorageEntryModifier::Optional => quote!(()),
     };
 
-    // Does the entry have a default value?
-    let is_defaultable_type = if has_default_value {
-        quote!(#crate_path::storage::address::Yes)
-    } else {
-        quote!(())
-    };
+    let all_fns = (0..=keys.len()).map(|n_keys| {
+        let keys_slice = &keys[..n_keys];
+        let (fn_name, is_fetchable, is_iterable) = if n_keys == keys.len() {
+            let fn_name = format_ident!("{snake_case_name}");
+            (fn_name, true, false)
+        } else {
+            let fn_name = if n_keys == 0 {
+                format_ident!("{snake_case_name}_iter")
+            } else {
+                format_ident!("{snake_case_name}_iter{}", n_keys)
+            };
+            (fn_name, false, true)
+        };
+        let is_fetchable_type = is_fetchable.then_some(quote!(#crate_path::storage::address::Yes)).unwrap_or(quote!(()));
+        let is_iterable_type = is_iterable.then_some(quote!(#crate_path::storage::address::Yes)).unwrap_or(quote!(()));
```

Comment:
nit: I found this a bit hard to follow, maybe a match would state the intent a bit better.
And since the `is_fetchable` and `is_iterable` are mutually exclusive, we could probably do with just one of them, up to you

Reply:
I had a go and came up with something like this:

```rust
let is_iterator = n_keys != keys.len();

let fn_name = match (is_iterator, keys.len()) {
    (false, _) => format_ident!("{snake_case_name}"),
    (true, 0) => format_ident!("{snake_case_name}_iter"),
    (true, n) => format_ident!("{snake_case_name}_iter{}", n)
};

let yes_ty = || quote!(#crate_path::storage::address::Yes);
let no_ty = || quote!(());

let is_fetchable_type = is_iterator.then(no_ty).unwrap_or_else(yes_ty);
let is_iterable_type = is_iterator.then(yes_ty).unwrap_or_else(no_ty);
```